File: android/app/src/main/kotlin/com/avrokeyboard/app/ime/voice/VoiceSession.kt
package com.avrokeyboard.app.ime.voice

enum class VoiceState {
    IDLE,
    REQUESTING_PERMISSION,
    LISTENING,
    RECOGNIZING,
    FINALIZING,
    ERROR,
}

/** In-memory voice compose. Partial text replaces, never concatenates. */
class VoiceSession {
    var state: VoiceState = VoiceState.IDLE
        private set
    var partial: String = ""
        private set
    var lastError: String? = null
        private set

    val isActive: Boolean
        get() = state == VoiceState.LISTENING ||
            state == VoiceState.RECOGNIZING ||
            state == VoiceState.REQUESTING_PERMISSION

    fun requestPermission() {
        state = VoiceState.REQUESTING_PERMISSION
        lastError = null
    }

    fun startListening() {
        state = VoiceState.LISTENING
        partial = ""
        lastError = null
    }

    fun onPartial(text: String) {
        if (state != VoiceState.LISTENING && state != VoiceState.RECOGNIZING) return
        if (text.isBlank()) return
        partial = text.trim()
        state = VoiceState.RECOGNIZING
    }

    fun onFinal(text: String): String? {
        state = VoiceState.FINALIZING
        val out = text.trim().ifEmpty { partial }.trim()
        partial = ""
        state = VoiceState.IDLE
        return out.ifEmpty { null }
    }

    fun cancel(): String? {
        val leftover = partial.trim().ifEmpty { null }
        partial = ""
        state = VoiceState.IDLE
        lastError = null
        return leftover
    }

    fun fail(code: String) {
        partial = ""
        lastError = code
        state = VoiceState.ERROR
    }

    fun recover() {
        if (state == VoiceState.ERROR) {
            state = VoiceState.IDLE
            lastError = null
        }
    }

    fun reset() {
        partial = ""
        lastError = null
        state = VoiceState.IDLE
    }
}

File: lib/ime/voice_session.dart
enum VoiceState {
  idle,
  requestingPermission,
  listening,
  recognizing,
  finalizing,
  error,
}

/// Mirrors Kotlin VoiceSession. Partial results replace, never append.
class VoiceSession {
  VoiceState state = VoiceState.idle;
  String partial = '';
  String? lastError;

  bool get isActive =>
      state == VoiceState.listening ||
      state == VoiceState.recognizing ||
      state == VoiceState.requestingPermission;

  void requestPermission() {
    state = VoiceState.requestingPermission;
    lastError = null;
  }

  void startListening() {
    state = VoiceState.listening;
    partial = '';
    lastError = null;
  }

  void onPartial(String text) {
    if (state != VoiceState.listening && state != VoiceState.recognizing) {
      return;
    }
    if (text.trim().isEmpty) return;
    partial = text.trim();
    state = VoiceState.recognizing;
  }

  String? onFinal(String text) {
    state = VoiceState.finalizing;
    final out = text.trim().isEmpty ? partial : text.trim();
    partial = '';
    state = VoiceState.idle;
    return out.isEmpty ? null : out;
  }

  String? cancel() {
    final leftover = partial.trim().isEmpty ? null : partial.trim();
    partial = '';
    state = VoiceState.idle;
    lastError = null;
    return leftover;
  }

  void fail(String code) {
    partial = '';
    lastError = code;
    state = VoiceState.error;
  }

  void recover() {
    if (state == VoiceState.error) {
      state = VoiceState.idle;
      lastError = null;
    }
  }

  void reset() {
    partial = '';
    lastError = null;
    state = VoiceState.idle;
  }
}
